File: extension/background_handlers/getAllTabs.js
/**
 * getAllTabs — background handler.
 *
 * 获取指定浏览器窗口（或所有窗口）中打开的全部标签页列表。
 * 结果写回变量走 runner 约定：返回 { value: [...], count }，runner 把 value（数组）写入 resultVar。
 *
 * 目标窗口解析：
 *   - windowVar 提供且解析为窗口对象（{windowId, tabId}）→ 只取该窗口的标签页；
 *   - 否则使用当前工作窗口（agent.workWindowId）；
 *   - 两者都没有 → 返回所有窗口的全部标签页（"所有打开的网页标签页"的兜底）。
 */
registerBackgroundHandler('getAllTabs', async function(step, agent) {
  const extra = step.extra || {};
  const onlyWebPages = extra.onlyWebPages !== false;

  let query = {};
  const wv = extra.windowVar;
  if (wv && typeof wv === 'object' && wv.windowId != null) {
    query.windowId = wv.windowId;
  } else if (agent.workWindowId) {
    query.windowId = agent.workWindowId;
  }
  // query 为空对象 → chrome.tabs.query({}) 返回所有窗口的所有标签页

  let tabs = await chrome.tabs.query(query);

  if (onlyWebPages) {
    tabs = tabs.filter(t => t.url && /^https?:/i.test(t.url) && !/^chrome-extension:/i.test(t.url));
  }

  const items = tabs.map(t => ({
    index: t.index,
    id: t.id,
    windowId: t.windowId,
    title: t.title || '',
    url: t.url || '',
    active: !!t.active,
    pinned: !!t.pinned,
    discarded: !!t.discarded,
  }));

  // 跨窗口按窗口 id、窗口内按 index 排序，保证结果确定、可预期
  items.sort((a, b) => (a.windowId - b.windowId) || (a.index - b.index));

  return { value: items, count: items.length };
});

File: src/runtime/commands/extension_commands/getAllTabs.py
"""Command: 获取所有标签页"""
from src.runtime.workflow.handlers.registry import register_handler, Param

@register_handler(cmd="getAllTabs", label="获取所有标签页",
    category="浏览器操作", runtime="extension",
    icon="fa-clone", icon_color="text-blue-500",
    bg_color="bg-blue-50",
    description="获取指定浏览器窗口（或所有窗口）中打开的全部标签页列表，可保存到变量",
    category_order=10,
    command_order=45,
)
class GetAllTabsHandler:
    params = [
        Param("windowVar", "浏览器窗口", "string", group="input", placeholder="如 browser1，留空使用当前工作窗口"),
        Param("onlyWebPages", "仅网页", "boolean", default=True, group="advanced", description="过滤 chrome://、edge://、扩展页、file://、about: 等受限页面"),
        Param("resultVar", "保存到变量", "string", default="tabs", group="output", description="标签页列表（数组，每项含 index/url/title/active/id/windowId）将保存到此变量"),
    ]